### DISCUSSION TITLE
Tabview does not update when scoping in a foreach loop

### DISCUSSION THREAD

**Discussion by Scocher** [pointfreeco/swift-composable-architecture#2850] [Q&A] [answered]
### Description

When running this code the updateIndex does not update.
`TabView(selection: $store.selectedID.sending(\.updateIndex)) {
        ForEach(Array(store.scope(state: \.rows, action: \.rows).enumerated()), id: \.element) { index, store in
          Text("\(store.id)").tag(store.id)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background((index % 2 == 0 ? Color.blue : Color.red).opacity(0.2))
        }
      }
      .tabViewStyle(PageTabViewStyle(indexDisplayMode: .always))`
compared to this:
`TabView(selection: $store.selectedID.sending(\.updateIndex)) {
        ForEach(Array(store.rows.enumerated()), id: \.element) { index, item in
          Text("\(item.id)").tag(item.id)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background((index % 2 == 0 ? Color.blue : Color.red).opacity(0.2))
        }
      }
      .tabViewStyle(PageTabViewStyle(indexDisplayMode: .always))`

https://github.com/pointfreeco/swift-composable-architecture/assets/2694492/bb95a754-23fa-4dc1-ab75-63f9b752eca1

[SimpleProject.zip](https://github.com/pointfreeco/swift-composable-architecture/files/14358297/SimpleProject.zip)

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

I expect the index to update.

### Actual behavior

It does not.

### Steps to reproduce

Run the project I have attached. 

### The Composable Architecture version information

'1.8.2', 'main'

### Destination operating system

iOS 17.2

### Xcode version information

15.2

### Swift Compiler version information

```shell
swift-driver version: 1.87.3 Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5)
Target: arm64-apple-macosx14.0
```

**Comment by domagojstankovic** [comment, marked as answer]
@Scocher seems like adding `id: \.state.id` to `ForEach` does the job, at least it worked for me.

[Link](https://pointfreecommunity.slack.com/archives/C04KQQ7NXHV/p1708131024728349?thread_ts=1707770625.548829&cid=C04KQQ7NXHV) to the original solution.

**Reply by Scocher** [reply]
Thanks for the reply. You are right about that I should use the state.id. I must have gotten lost in the enumerated code. ~~Unfortunately, this did not solved the issue. But I will look into the link and see if I can find a solution there.~~

**Edit: Sorry. You were right. That was the problem. I have to read my code more carefully.**

**Comment by stephencelis** [comment]
@domagojstankovic Notes the issue, which is that the identity should be based on the data in the store and not the identity of the store itself. I'm going to convert this to a discussion since it's not technically a bug with the library.
